mirgecom/gas_model.py

```diff
@@ -743,3 +743,115 @@ def replace_fluid_state(
+            discr, dd_vol, dd_faces, state.temperature)
+        gamma = volume_and_surface_quadrature_interpolation(
+            discr, dd_vol, dd_faces, gamma)
+
+    # Convert back to conserved varaibles and use to make the new fluid state
+    cv_modified = entropy_to_conservative_vars(gamma, ev_quad)
+
+    return make_fluid_state(cv=cv_modified,
+                            gas_model=gas_model,
+                            temperature_seed=temperature_seed)
+
+
+def conservative_to_entropy_vars(gamma, state):
+    """Compute the entropy variables from conserved variables.
+
+    Converts from conserved variables (density, momentum, total energy)
+    into entropy variables.
+
+    Parameters
+    ----------
+    state: :class:`~mirgecom.gas_model.FluidState`
+        The full fluid conserved and thermal state
+
+    Returns
+    -------
+    ConservedVars
+        The entropy variables
+    """
+    from mirgecom.fluid import make_conserved
+
+    dim = state.dim
+    actx = state.array_context
+
+    rho = state.mass_density
+    u = state.velocity
+    p = state.pressure
+    rho_species = state.species_mass_density
+
+    u_square = sum(v ** 2 for v in u)
+    s = actx.np.log(p) - gamma*actx.np.log(rho)
+    rho_p = rho / p
+    rho_species_p = rho_species / p
+
+    return make_conserved(
+        dim,
+        mass=((gamma - s)/(gamma - 1)) - 0.5 * rho_p * u_square,
+        energy=-rho_p,
+        momentum=rho_p * u,
+        species_mass=((gamma - s)/(gamma - 1)) - 0.5 * rho_species_p * u_square
+    )
+
+
+def entropy_to_conservative_vars(gamma, ev: ConservedVars):
+    """Compute the conserved variables from entropy variables *ev*.
+
+    Converts from entropy variables into conserved variables
+    (density, momentum, total energy).
+
+    Parameters
+    ----------
+    ev: ConservedVars
+        The entropy variables
+
+    Returns
+    -------
+    ConservedVars
+        The fluid conserved variables
+    """
+    from mirgecom.fluid import make_conserved
+
+    dim = ev.dim
+    actx = ev.array_context
+    # See Hughes, Franca, Mallet (1986) A new finite element
+    # formulation for CFD: (DOI: 10.1016/0045-7825(86)90127-1)
+    inv_gamma_minus_one = 1/(gamma - 1)
+
+    # Convert to entropy `-rho * s` used by Hughes, France, Mallet (1986)
+    ev_state = ev * (gamma - 1)
+    v1 = ev_state.mass
+    v234 = ev_state.momentum
+    v5 = ev_state.energy
+    v6ns = ev_state.species_mass
+
+    v_square = sum(v**2 for v in v234)
+    s = gamma - v1 + v_square/(2*v5)
+    s_species = gamma - v6ns + v_square/(2*v5)
+    iota = ((gamma - 1) / (-v5)**gamma)**(inv_gamma_minus_one)
+    rho_iota = iota * actx.np.exp(-s * inv_gamma_minus_one)
+    rho_iota_species = iota * actx.np.exp(-s_species * inv_gamma_minus_one)
+
+    return make_conserved(
+        dim,
+        mass=-rho_iota * v5,
+        energy=rho_iota * (1 - v_square/(2*v5)),
+        momentum=rho_iota * v234,
+        species_mass=-rho_iota_species * v5
+    )
```

Comment:
Do these do something different vs. the EOS versions?

Reply:
The EOS versions are currently unused.  We may switch to those.. not sure yet.